### PR TITLE
Ek checkpoint

### LINKED ISSUE #2489: Implement Py interface for EK checkpointing + Test
Core code seems to be in src/core/grid_based_algorithms/elekctrokinetics_pdb_parser.cpp

### PR DESCRIPTION
**Pull request by itischler**
Fixes #2489 

Description of changes:
 - enabled ek checkpointing
 - species are now a part of ``ek._params``

Checkpointing does not work with thermostat (see #2652)

PR Checklist
------------
 - [ ] Tests?
   - [ ] Interface
   - [ ] Core 
 - [ ] Docs?

### CODE REVIEW

File: doc/sphinx/advanced_methods.rst [hunk 1]
```diff
@@ -1562,6 +1562,22 @@ the options associated to these shapes. In order to properly set up the
...
 
+.. _Checkpointing:
+
+Checkpointing
+^^^^^^^^^^^^^
+::
+
+    ek.save_checkpoint(path)
+
+Checkpointing in the EK works quite similar to checkpointing in the LB, because the density in not saved within the :class:`espressomd.checkpointing` object. However one should keep in mind, that the EK not only saved the density of the species but also saves the population of the LB fluid in a seperate file. To load a checkpoint the :class:`espressomd.electrokinetics.Electrokinetics` should have the same name as in the scribt it was saved, but to use the species one need to extract them from the :class:`espressomd.electrokinetics.Electrokinetics` via ``species'``.
```
Comment (fweik) [inline doc/sphinx/advanced_methods.rst]: Please check check the spelling of the paragraph (e.g. `scribt`) :-)

File: src/python/espressomd/electrokinetics.pyx [hunk 2]
```diff
@@ -311,10 +315,17 @@ IF ELECTROKINETICS:
             else:
                 raise Exception("'es_coupling' is not active.")
 
-        # TODO:
-        def checkpoint(self):
-            raise Exception(
-                "Please implement this method in the pickle routine.")
+        def save_checkpoint(self, path):
```
Comment (fweik) [inline src/python/espressomd/electrokinetics.pyx:320]: Please use the python [tempfile module](https://docs.python.org/2/library/tempfile.html) to generate temporary files.
That should be more portable and check that the file can be created and so on.

**Review by fweik — CHANGES_REQUESTED** [review]
Please also add in the documentation that species can now be created using a parameter (this is actually convenient because it allow creating the complete ek in one line.).

**Comment by fweik** [comment]
@itischler what is the status here?

**Comment by itischler** [comment]
Did not have the time to work on the changes yet.

File: src/core/grid_based_algorithms/electrokinetics_cuda.cu [hunk 3]
```diff
@@ -4045,10 +4045,9 @@ int ek_neutralize_system(int species) {
   return 0;
 }
 
-int ek_save_checkpoint(char *filename) {
+int ek_save_checkpoint(char *filename, char *lb_filename) {
   std::string fname(filename);
-  std::ofstream fout((const char *)(fname + ".ek").c_str(),
-                     std::ofstream::binary);
+  std::ofstream fout((const char *)(fname).c_str(), std::ofstream::binary);
```
Comment (fweik) [inline src/core/grid_based_algorithms/electrokinetics_cuda.cu]: This look odd to me, why is this cast needed?

File: src/core/grid_based_algorithms/electrokinetics_cuda.cu [hunk 4]
```diff
@@ -4068,7 +4067,8 @@ int ek_save_checkpoint(char *filename) {
   free(densities);
   fout.close();
 
-  lb_lbfluid_save_checkpoint((char *)(fname + ".lb").c_str(), 1);
+  std::string lb_fname(lb_filename);
+  lb_lbfluid_save_checkpoint((char *)(lb_fname).c_str(), 1);
```
Comment (fweik) [inline src/core/grid_based_algorithms/electrokinetics_cuda.cu]: same here

File: src/core/grid_based_algorithms/electrokinetics_cuda.cu (hunk 3 above)
Comment (itischler) [reply src/core/grid_based_algorithms/electrokinetics_cuda.cu]: it is not needed and I removed it

**Review by fweik — APPROVED** [review]

**Comment by fweik** [comment]
bors retry

**Pull request merged**